File: python/mandatekit/canonical.py
"""
Canonical JSON for signing.

A mandate is signed over its canonical byte form so that any party can
re-serialize the same mandate and get the same bytes, hence the same signature
check. v0 uses sorted keys + compact separators. This is deliberately simple and
documented; a 1.0 would adopt RFC 8785 (JCS) for full number-canonicalization
guarantees. The TypeScript port produces byte-identical output.
"""

import json


def canonicalize(obj) -> bytes:
    return json.dumps(
        obj, sort_keys=True, separators=(",", ":"), ensure_ascii=False
    ).encode("utf-8")

File: typescript/src/canonical.ts
/**
 * Canonical JSON for signing. Sorted keys + compact separators, byte-identical
 * to the Python port's `canonicalize`. See the Python module for the rationale
 * and the v0-vs-JCS note.
 */

function sortValue(v: unknown): unknown {
  if (Array.isArray(v)) return v.map(sortValue);
  if (v && typeof v === "object") {
    const out: Record<string, unknown> = {};
    for (const k of Object.keys(v as Record<string, unknown>).sort()) {
      out[k] = sortValue((v as Record<string, unknown>)[k]);
    }
    return out;
  }
  return v;
}

export function canonicalize(obj: unknown): Buffer {
  // JSON.stringify uses no whitespace and ',' / ':' separators by default,
  // matching Python's compact separators.
  return Buffer.from(JSON.stringify(sortValue(obj)), "utf8");
}
